src/devices/BuildHat/Brick.cs

```diff
@@ -0,0 +1,1564 @@
+        {
+            PortWrite($"port {(byte)port} ; plimit 1 ; off\r");
+        }
+
+        /// <summary>
+        /// Gets the sensor connected.
+        /// </summary>
+        /// <param name="port">The port.</param>
+        /// <returns>A sensor.</returns>
+        /// <exception cref="Exception">A motor or no sensor is connected</exception>
+        public Sensor GetSensor(SensorPort port)
+        {
+            return (Sensor)_elements[(byte)port];
+        }
+
+        /// <summary>
+        /// Gets the sensor connected.
+        /// </summary>
+        /// <param name="port">The port.</param>
+        /// <returns>A sensor.</returns>
+        /// <exception cref="Exception">A motor or no sensor is connected</exception>
+        public ActiveSensor GetActiveSensor(SensorPort port)
+        {
+            if (!IsActiveSensor(_sensorType[(byte)port]))
+            {
+                throw new Exception("A motor or no sensor is connected");
+            }
+
+            return (ActiveSensor)_elements[(byte)port];
+        }
+
+        /// <summary>
+        /// Writes directly to the sensor. The bytes to the current port, the first one or two bytes being header bytes. The message is padded if
+        /// necessary, and length and checksum fields are automatically populated.
+        /// </summary>
+        /// <param name="port">The port.</param>
+        /// <param name="data">The buffer to send.</param>
+        /// <param name="singleHeader">True for single header byte.</param>
+        public void WriteBytesToSensor(SensorPort port, ReadOnlySpan<byte> data, bool singleHeader)
+        {
+            StringBuilder command = new StringBuilder();
+            command.Append(singleHeader ? "write1 " : "write2 ");
+            command.Append(string.Join(" ", data.ToArray().Select(m => string.Format("{0:X2} ", m))));
+            command.Append("\r");
+            PortWrite(command.ToString());
+        }
+
+        #endregion
+
+        /// <summary>
+        /// Send a raw command. This can be used for specific sensors or setup sensors.
+        /// </summary>
+        /// <param name="command">The command.</param>
+        public void SendRawCommand(string command)
+        {
+            if (string.IsNullOrEmpty(command))
+            {
+                return;
+            }
+
+            if (!command.EndsWith("\r"))
+            {
+                command += "\r";
+            }
+
+            PortWrite(command);
+        }
+
+        /// <summary>
+        /// Wait to get a sensor connected on a specific port.
+        /// </summary>
+        /// <param name="port">The port.</param>
+        /// <param name="token">A cancellatin token</param>
+        public void WaitForSensorToConnect(SensorPort port, CancellationToken token = default)
+        {
+            // Test sensor, plug a RGB distance one on port C
+            while (GetSensor(port) == null)
+            {
+                token.WaitHandle.WaitOne(50);
```

Comment:
That will never terminate if no sensor is plugged in. Test the cancellation token in the loop.

Reply:
yes, good catch